force_bdss/data_sources/base_data_source_model.py

```diff
@@ -68,6 +68,50 @@ def _trigger_changes_slots(self, obj, name, new):
         if changes_slots:
             self.changes_slots = True
 
+    @on_trait_change('changes_slots')
+    def _update_slot_info(self):
+        """This method is designed to be performed upon a `change_slots`
+        event to obtain the new format for input_slot_info
+        and output_slot_info lists. It obtains the new input_slot_info
+        and output_slot_info defaults, and updates the existing
+        attributes with any format changes.
+
+        A change_slots event can indicate 2 outcomes:
+        1. Changes to the number of input and output Slot objects
+        returned by the corresponding BaseDataSource
+        2. Changes to attributes on each Slot object returned by the
+        corresponding BaseDataSource
+
+        In the first instance, we would like to retain any InputSlotInfo
+        or OutputSlotInfo elements referring to variables that have
+        been defined before. These are identified as having matching
+        `type` and `description` attributes with an element in the
+        new default slot lists.
+
+        In the second instance, expect each element on the existing
+        input_slot_info or output_slot_info lists to refer to the
+        corresponding element on the new default lists returned by
+        the _return_slots method. Therefore we simply update any
+        attribute `type` and `description` values.
+        """
+        # Get new slots, caused by change_slots event
+        for attr_name, slot_info in self._slot_info_generator():
+            attr = getattr(self, attr_name)
+
+            if len(slot_info) == len(attr):
+                # If changes_slots event has not altered the length of
+                # each list attribute, update each element with any
+                # changed 'type' and 'description' values
+                merge_lists(slot_info, attr, ['type', 'description'])
+            else:
+                # Otherwise update the list attributes by retaining any
+                # elements that have matching 'type' and 'description'
+                # values
+                new_list = retain_list(
+                    slot_info, attr, ['type', 'description']
+                )
+                setattr(self, attr_name, new_list)
```

Comment:
This is more of a comment on the `retain_list` and `merge_lists` behaviour: in one case we `setattr` implicitly, while for the other one we have to do it manually. Do I miss the design pattern here, or it could be possible to unify these two slot updates?

Reply:
Yes, I agree this looks a bit strange. The `retain_list` function returns a new list that contains any elements of the old list (`attr` argument) that we want to keep.

A different design of `retain_list` would rather mutate `attr` in place, however since we are inserting or removing elements, this becomes difficult to do whilst iterating through them.

Reply:
What happens if instead of `merge_lists(slot_info, attr, ['type', 'description'])` we use
```
new_list = retain_list(
                    slot_info, attr, ['type', 'description']
                )
                setattr(self, attr_name, new_list)
```
?

Reply:
Ah, yes we could do - actually that would be a more robust way since the `merge_lists` function assumes that the order of elements in each list is the same (they are referring to the same variables)